src/github.com/getlantern/proxiedsites/proxiedsites.go

```diff
@@ -1,269 +1,218 @@
-		Additions: additions,
-		Deletions: set.StringSlice(delSet),
-	}
+// equals checks whether this configsets is identical to some other configsets
+func (cs *configsets) equals(other *configsets) bool {
+	return cs.cloud.IsEqual(other.cloud) &&
+		cs.add.IsEqual(other.add) &&
+		cs.del.IsEqual(other.del)
 }
 
-// Update modifies an existing ProxiedSites instance
-// to include new addition and deletion deltas sent from
-// the client
-func (ps *ProxiedSites) Update(cfg *Config) {
-
-	for i := range cfg.Additions {
-		log.Debugf("Adding site %s", cfg.Additions[i])
-		ps.addSet.Add(cfg.Additions[i])
-		// remove any new sites from our deletions list
-		// if they were previously added there
-		ps.delSet.Remove(cfg.Additions[i])
+// Configure applies the given configuration. If there were changes, a Delta is
+// returned that includes the additions and deletions from the active list. If
+// there were no changes, or the changes couldn't be applied, this method
+// returns a nil Delta.
+func Configure(cfg *Config) *Delta {
+	newCS := cfg.toCS()
+	if cs != nil && cs.equals(newCS) {
+		log.Debug("Configuration unchanged")
+		return nil
 	}
 
-	for i := range cfg.Deletions {
+	newPacFile, err := generatePACFile(newCS.activeList)
+	if err != nil {
+		log.Errorf("Error generating pac file, leaving configuration unchanged: %v", err)
+		return nil
+	}
 
-		if ps.addSet.Has(cfg.Deletions[i]) {
-			// if a new deletion was previously on our
-			// additionss list, remove it here
-			ps.addSet.Remove(cfg.Deletions[i])
+	var delta *Delta
+	if cs == nil {
+		delta = &Delta{
+			Additions: newCS.activeList,
 		}
-		if ps.cloudSet.Has(cfg.Deletions[i]) {
-			// add to the delete list only if it's
-			// already in the global list
-			ps.delSet.Add(cfg.Deletions[i])
+	} else {
+		delta = &Delta{
+			Additions: toStrings(set.Difference(newCS.active, cs.active)),
+			Deletions: toStrings(set.Difference(cs.active, newCS.active)),
 		}
 	}
-
-	ps.cfg.Deletions = set.StringSlice(ps.delSet)
-	ps.cfg.Additions = set.StringSlice(set.Difference(ps.addSet, ps.cloudSet))
-
-	ps.entries = set.StringSlice(set.Difference(set.Union(ps.cloudSet, ps.addSet),
-		ps.delSet))
-	go ps.updatePacFile()
-}
-
-func (ps *ProxiedSites) GetConfig() *Config {
-	return ps.cfg
+	cs = newCS
+	pacFile = newPacFile
+	log.Debug("Applied updated configuration")
+	return delta
 }
 
-func GetPacFile() string {
-	return PacFilePath
+// ActiveDelta returns the active sites as a Delta of additions.
+func ActiveDelta() *Delta {
+	cfgMutex.RLock()
+	d := &Delta{
+		Additions: cs.activeList,
+	}
+	cfgMutex.RUnlock()
+	return d
 }
 
-func SetPacFile(pacFile string) {
-	PacFilePath = pacFile
+// ServePAC serves up the PAC file and can be used as an http.HandlerFunc
+func ServePAC(resp http.ResponseWriter, req *http.Request) {
+	resp.Header().Set("Content-Type", "application/x-ns-proxy-autoconfig")
+	resp.WriteHeader(http.StatusOK)
+	cfgMutex.RLock()
+	resp.Write([]byte(pacFile))
+	cfgMutex.RUnlock()
```

Comment:
Nice. Definitely makes sense to store it in-memory and skip saving a file entirely since we already write the latest changes to the YAML configuration.